File: rotate-array-k-times.py
class Solution:
    def rotate(self, nums: List[int], k: int) -> None:
        """
        Do not return anything, modify nums in-place instead.
        """
        if len(nums) == 0:
            return

        def reverse(nums, s, e):
            j = 0
            for i in range(s, ((s + e) // 2) + 1):
                temp = nums[e - j]
                nums[e - j] = nums[i]
                nums[i] = temp
                j += 1
                # print('num',nums)
            return nums

        if k >= len(nums):
            k = k % len(nums)  # rotating array by 5 time in a arr of length of 5 is same as rotating by 1
        reverse(nums, 0, len(nums) - 1)  # o(n)
        reverse(nums, 0, k - 1)
        reverse(nums, k, len(nums) - 1)  # O(n)

        # time-o(n) space O(1)

File: RotateArrayKTimes.java
class Solution {
    //(O(1) space and O(n) time
    public void rotate(int[] nums, int k) {

        if (nums == null || nums.length == 0) return;
        int n = nums.length;
        if (k >= n) k = k%n;
        //Right Rotate
        reverse(nums, 0, n - 1);
        reverse(nums, 0, k - 1);
        reverse(nums, k, n - 1);

        //Left Rotate
        //reverse(nums, 0, n - 1);
        //reverse(nums, 0, n - k - 1);
        //reverse(nums, n - k, n - 1);

    }

    private void reverse(int[] nums, int start, int end) {

        while (start < end) {

            swap(nums, start, end);
            start++; end--;

        }

    }

    private void swap(int[] nums, int start, int end) {

        int temp = nums[start];
        nums[start] = nums[end];
        nums[end] = temp;
    }
}